src/dbnode/encoding/proto/encoder.go

```diff
@@ -670,77 +666,71 @@ func (enc *Encoder) encodeProtoValues(m *dynamic.Message) error {
 
 	// Reset for re-use.
 	enc.changedValues = enc.changedValues[:0]
-	changedFields := enc.changedValues
-
 	enc.fieldsChangedToDefault = enc.fieldsChangedToDefault[:0]
-	fieldsChangedToDefault := enc.fieldsChangedToDefault
-
-	if enc.lastEncoded == nil {
-		enc.lastEncoded = dynamic.NewMessage(enc.schema)
-	}
-
-	for _, fieldNum := range enc.protoFields {
-		var (
-			field       = enc.schema.FindFieldByNumber(fieldNum)
-			fieldNumInt = int(fieldNum)
-			prevVal     = enc.lastEncoded.GetFieldByNumber(fieldNumInt)
-			curVal      = m.GetFieldByNumber(fieldNumInt)
-		)
 
-		if fieldsEqual(curVal, prevVal) {
-			// Clear fields that haven't changed.
-			if err := m.TryClearFieldByNumber(fieldNumInt); err != nil {
-				return fmt.Errorf("error: %v clearing field: %d", err, fieldNumInt)
-			}
-		} else {
-			isDefaultValue, err := isDefaultValue(field, curVal)
-			if err != nil {
-				return fmt.Errorf(
-					"error: %v, checking if %v is default value for field %s",
-					err, curVal, field.String())
-			}
-			if isDefaultValue {
-				fieldsChangedToDefault = append(fieldsChangedToDefault, fieldNum)
+	currProtoFields := enc.unmarshaler.nonCustomFieldValues()
+	for i, protoField := range enc.protoFields {
+		var curVal []byte
+		for _, val := range currProtoFields {
+			if protoField.fieldNum == val.fieldNum {
+				curVal = val.marshaled
 			}
+		}
 
-			changedFields = append(changedFields, fieldNum)
-			if err := enc.lastEncoded.TrySetFieldByNumber(fieldNumInt, curVal); err != nil {
-				return fmt.Errorf(
-					"error: %v setting field %d with value %v on lastEncoded",
-					err, fieldNumInt, curVal)
-			}
+		prevVal := protoField.marshaled
+		if bytes.Equal(prevVal, curVal) {
+			// No change, nothing to encode.
+			continue
 		}
+
+		if curVal == nil {
+			// Interpret as default value.
+			enc.fieldsChangedToDefault = append(enc.fieldsChangedToDefault, protoField.fieldNum)
+		}
+
+		enc.changedValues = append(enc.changedValues, protoField.fieldNum)
+		// Need to copy since the encoder no longer owns the original source of the bytes once
+		// this function returns.
+		enc.protoFields[i].marshaled = append(enc.protoFields[i].marshaled[:0], curVal...)
 	}
 
-	if len(changedFields) == 0 {
+	if len(enc.changedValues) == 0 {
 		// Only want to skip encoding if nothing has changed AND we've already
 		// encoded the first message.
 		enc.stream.WriteBit(opCodeNoChange)
 		return nil
 	}
 
-	marshaled, err := m.MarshalAppend(enc.marshalBuf[:0])
-	if err != nil {
-		return fmt.Errorf("%s error trying to marshal protobuf: %v", encErrPrefix, err)
+	marshalBuf := enc.marshalBuf[:0]
```

Comment:
Hm, seems you reset the `enc.marshalBuf` back to the value of this local var.  Perhaps remove the `marshalBuf` var altogether and just use `enc.marshalBuf` below?